sympy/core/operations.py

```diff
@@ -504,3 +522,160 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher:
+    """
+    Handler dispatcher for associative operators
+
+    .. notes::
+       This approach is experimental, and can be replaced or deleted in the future.
+       See https://github.com/sympy/sympy/pull/19463.
+
+    Explanation
+    ===========
+
+    If arguments of different kind are passed, their ``_op_priority`` are compared to
+    choose the type with the highest priority. If this is ambiguous, priority relations
+    which are registered by ``register_priority`` method are referred to.
+    Once a type is selected, handler method of the type, which is a static method, is used
+    to perform the operation.
+
+    Priority registeration is unordered. You cannot make ``A*B`` and ``B*A`` use different
+    handler method. Every logic dealing with the order of arguments must be implemented in
+    the handler method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.add import add
+
+    >>> class NewExpr(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @add.register_priority(Expr, NewExpr)
+    ... @add.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     return NewExpr
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> add(a, b) == NewAdd(a, b)
+    True
+
+    Registered priority selector can return any type.
+
+    >>> class Foo(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class Bar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class FooBar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+
+    >>> @add.register_priority(Foo, Bar)
+    ... def _(_1, _2):
+    ...     return FooBar
+
+    >>> add(Foo(), Bar()) == NewAdd(Foo(), Bar())
+    True
+
+    """
+    def __init__(self, name, doc=None):
+        self.name = name
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self._handlergetter = attrgetter(self.handlermethod)
+        self._dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
```

Comment:
`Dispatcher.__doc__` also exists. Can you add something similar here that prints out the registered priority handlers?

Reply:
Maybe it would be better to implement `AssocOpDispatcher.__doc__`.

Reply:
Sorry, that's what I mean :)

Reply:
Done in 3a065cb.

```python
>>> from sympy.core.add import add
>>> print(add.__doc__)
Multiply dispatched method: add

Registered priority relations are as follows:

Inputs: <Expr, MatrixExpr>
---------------------------
Use MatrixExpr's handler method.

Inputs: <MatrixExpr, Expr>
---------------------------
Use MatrixExpr's handler method.

Inputs: <MatrixExpr, MatrixExpr>
---------------------------------
Use MatrixExpr's handler method.
```